File: java/datastructures/fenwicktree.java
import java.util.*;
import java.io.*;
// this fenwick tree is for range update and point query
class FenwickTree{
  private int[] ft;
  public FenwickTree(int n){
    ft = new int[n];
  }
  private int LSOne(int a){
    return (a&(-a));
  }
  private void adjust(int k,int value){
    for(;k<ft.length;k+=LSOne(k))ft[k]+=value;
  }
  public void range_update(int l,int r,int value){
    this.adjust(l,value);
    this.adjust(r+1,-1*value);
  }
  public int rsq(int a){
    int ans=0;
    for(;a>0;a-=LSOne(a)){
      ans+=ft[a];
    }
    return ans;
  }
}
class fenwicktree{
  public static void main(String args[])throws IOException{
    BufferedReader r =  new BufferedReader(new InputStreamReader(System.in));
    int n = Integer.parseInt(r.readLine());
    FenwickTree f = new FenwickTree(n);
    for(int i=0;i<n;i++){
      f.range_update(i+1,i+1,Integer.parseInt(r.readLine()));
    }
    System.out.println(f.rsq(2));
    f.range_update(1,5,10);
    System.out.println(f.rsq(2));
    System.out.println(f.rsq(8));
  }
}

File: c++/datastructures/fenwicktree.cpp
#include <bits/stdc++.h>
using namespace std;
class FenwickTree{
  private:
    vector<int> ft;
  public:
    FenwickTree(int n){
      ft.assign(n+1,0);
    }
    int LSOne(int a){return (a&-a);}
    int rsq(int a){
      int ans=0;
      for(;a;a-=LSOne(a))ans+=ft[a];
      return ans;
    }
    int rsq(int a,int b){
      if(a!=1){
        return rsq(b)-rsq(a-1);
      }else{
        return rsq(b);
      }
    }
    void adjust(int k,int value){
      for(;k<ft.size();k+=LSOne(k))ft[k]+=value;
    }
};
int main(){
  int n;
  cin>>n;
  int a[n];
  for(int i=0;i<n;i++){
    cin>>a[i];
  }
  FenwickTree f =FenwickTree(n);
  for(int i=0;i<n;i++){
    f.adjust(i+1,a[i]);
  }
  cout<<f.rsq(2,4)<<endl;
  f.adjust(2,10);
  cout<<f.rsq(2,4)<<endl;

  /*
  * for range update and point query adjust value to l and 
  * -value to r+1, so the point will give the correct value
  * no change in query function.
  */
  /*
  * for range update and range query two arrays are used.
  */
}